pkg/templateservicebroker/brokerapi/templateservicebroker/provision.go

```diff
@@ -0,0 +1,186 @@
+}
+
+func (b *Broker) ensureTP(namespace string, instance_id string, template *templateapi.Template, secret *kapi.Secret, didWork *bool) (*api.TemplateProvision, *brokerapi.Response) {
+	tp := &api.TemplateProvision{
+		ObjectMeta: kapi.ObjectMeta{Name: instance_id},
+		Spec: api.TemplateProvisionSpec{
+			Template:  *template,
+			SecretRef: kapi.LocalObjectReference{Name: secret.Name},
+		},
+		Status: api.TemplateProvisionStatus{
+			State: api.TemplateProvisionStateNew,
+		},
+	}
+
+	createdTP, err := b.oc.TemplateProvisions(namespace).Create(tp)
+	if err == nil {
+		*didWork = true
+		return createdTP, nil
+	}
+
+	if errors.IsAlreadyExists(err) {
+		var existingTP *api.TemplateProvision
+		existingTP, err = b.oc.TemplateProvisions(namespace).Get(tp.Name)
+		if err == nil && reflect.DeepEqual(tp.Spec, existingTP.Spec) {
+			return existingTP, nil
+		}
+
+		return nil, &brokerapi.Response{http.StatusConflict, brokerapi.ProvisionResponse{}, nil}
+	}
+
+	return nil, &brokerapi.Response{http.StatusInternalServerError, nil, err}
+}
+
+func (b *Broker) ensureSecretOwnerRef(secret *kapi.Secret, tp *api.TemplateProvision, didWork *bool) (*kapi.Secret, *brokerapi.Response) {
+	ownerRef := kapi.OwnerReference{
+		APIVersion: "v1",
+		Kind:       "TemplateProvision",
+		Name:       tp.Name,
+		UID:        tp.UID,
+	}
+	for _, ref := range secret.OwnerReferences {
+		if ref == ownerRef {
+			return secret, nil
+		}
+	}
+
+	secret.OwnerReferences = append(secret.OwnerReferences, ownerRef)
+	secret, err := b.kc.Secrets(secret.Namespace).Update(secret)
+	if err == nil {
+		*didWork = true
+		return secret, nil
+	}
+
+	return nil, &brokerapi.Response{http.StatusInternalServerError, nil, err}
+}
+
+func (b *Broker) ensureUUIDMap(instance_id string, tp *api.TemplateProvision, didWork *bool) (*api.UUIDMap, *brokerapi.Response) {
+	uuidMap := &api.UUIDMap{
+		ObjectMeta: kapi.ObjectMeta{Name: instance_id},
+		Spec: api.UUIDMapSpec{
+			TemplateProvisionRef: kapi.ObjectReference{
+				Kind:      tp.Kind,
+				Namespace: tp.Namespace,
+				Name:      tp.Name,
+			},
+		},
+	}
+
+	newUUIDMap, err := b.oc.UUIDMaps().Create(uuidMap)
+	if err == nil {
+		*didWork = true
+		return newUUIDMap, nil
+	}
+
+	if errors.IsAlreadyExists(err) {
+		var existingUM *api.UUIDMap
+		existingUM, err = b.oc.UUIDMaps().Get(uuidMap.Name)
+		if err == nil && reflect.DeepEqual(uuidMap.Spec, existingUM.Spec) {
+			return existingUM, nil
+		}
+
+		return nil, &brokerapi.Response{http.StatusConflict, brokerapi.ProvisionResponse{}, nil}
+	}
+
+	return nil, &brokerapi.Response{http.StatusInternalServerError, nil, err}
+}
+
+func (b *Broker) Provision(instance_id string, preq *brokerapi.ProvisionRequest) *brokerapi.Response {
+	namespace := preq.Parameters["namespace"]
+
+	template, err := b.templateFromUID(types.UID(preq.ServiceID))
+	if err != nil {
+		return &brokerapi.Response{http.StatusBadRequest, nil, err}
+	}
+
+	template.ObjectLabels = make(map[string]string)
```

Comment:
you should only do this if ObjectLabels is nil.  otherwise you're overwriting existing labels the template wants to set.